pkg/kubelet/images/pullmanager/fs_pullrecords.go

```diff
@@ -0,0 +1,275 @@
+	intentRecordPath := filepath.Join(f.pullingDir, cacheFilename(image))
+	if err := os.Remove(intentRecordPath); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+func (f *fsPullRecordsAccessor) GetImagePulledRecord(imageRef string) (*kubeletconfiginternal.ImagePulledRecord, bool, error) {
+	recordBytes, err := os.ReadFile(filepath.Join(f.pulledDir, cacheFilename(imageRef)))
+	if os.IsNotExist(err) {
+		return nil, false, nil
+	} else if err != nil {
+		return nil, false, err
+	}
+
+	pulledRecord, err := decodePulledRecord(f.decoder, recordBytes)
+	return pulledRecord, true, err
+}
+
+func (f *fsPullRecordsAccessor) ListImagePulledRecords() ([]*kubeletconfiginternal.ImagePulledRecord, error) {
+	var pullRecords []*kubeletconfiginternal.ImagePulledRecord
+	err := processDirFiles(f.pulledDir,
+		func(filePath string, fileContent []byte) error {
+			pullRecord, err := decodePulledRecord(f.decoder, fileContent)
+			if err != nil {
+				klog.V(5).InfoS("failed to deserialize, skipping file", "filePath", filePath, "error", err)
+				return nil
+			}
+			pullRecords = append(pullRecords, pullRecord)
+			return nil
+		})
+
+	return pullRecords, err
+}
+
+func (f *fsPullRecordsAccessor) WriteImagePulledRecord(pulledRecord *kubeletconfiginternal.ImagePulledRecord) error {
+	recordBytes := bytes.NewBuffer([]byte{})
+	if err := f.encoder.Encode(pulledRecord, recordBytes); err != nil {
+		return fmt.Errorf("failed to serialize ImagePulledRecord: %w", err)
+	}
+
+	return writeFile(f.pulledDir, cacheFilename(pulledRecord.ImageRef), recordBytes.Bytes())
+}
+
+func (f *fsPullRecordsAccessor) DeleteImagePulledRecord(imageRef string) error {
+	err := os.Remove(filepath.Join(f.pulledDir, cacheFilename(imageRef)))
+	if os.IsNotExist(err) {
+		return nil
+	}
+	return err
+}
+
+func cacheFilename(image string) string {
+	return fmt.Sprintf("sha256-%x", sha256.Sum256([]byte(image)))
+}
+
+// writeFile writes `content` to the file with name `filename` in directory `dir`.
+// It assures write atomicity by creating a temporary file first and only after
+// a successful write, it move the temp file in place of the target.
+func writeFile(dir, filename string, content []byte) error {
+	// create target folder if it does not exists yet
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return fmt.Errorf("failed to create directory %q: %w", dir, err)
+	}
+
+	targetPath := filepath.Join(dir, filename)
+	tmpPath := targetPath + ".tmp"
+	if err := os.WriteFile(tmpPath, content, 0600); err != nil {
+		return fmt.Errorf("failed to create temporary file %q: %w", tmpPath, err)
+	}
+
+	return os.Rename(tmpPath, targetPath)
+}
+
+// processDirFiles reads files in a given directory and peforms `fileAction` action on those.
+func processDirFiles(dirName string, fileAction func(filePath string, fileContent []byte) error) error {
+	return filepath.WalkDir(dirName, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
```

Comment:
I *think* returning err here short-circuits all file reading, which is not what we want

I think we should aggregate errors at the top level of processDirFiles, but keep walking the directory to read as many files as we can

```go
var errors []error
walkErr := filepath.WalkDir(dirName, func(path string, d fs.DirEntry, err error) error {
  if err != nil {
    errors = append(errors, err)
    return nil
  }
  ...
})

if walkErr != nil {
  errors = append(errors, walkErr)
}
return errors.NewAggregate(errors)
```